lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,648 @@
+use crate::BytesDecoder;
+
+use super::BoxedFramingError;
+use bytes::{Buf, Bytes, BytesMut};
+use derivative::Derivative;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+use tokio;
+use tokio::task::JoinHandle;
+use tokio_util::codec::Decoder;
+use tracing::{info, warn};
+use vector_config::configurable_component;
+
+const GELF_MAGIC: [u8; 2] = [0x1e, 0x0f];
+const GELF_MAX_TOTAL_CHUNKS: u8 = 128;
+const DEFAULT_TIMEOUT_MILLIS: u64 = 5000;
+const DEFAULT_PENDING_MESSAGES_LIMIT: usize = 1000;
+
+const fn default_timeout_millis() -> u64 {
+    DEFAULT_TIMEOUT_MILLIS
+}
+
+const fn default_pending_messages_limit() -> usize {
+    DEFAULT_PENDING_MESSAGES_LIMIT
+}
+
+/// Config used to build a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct ChunkedGelfDecoderConfig {
+    /// Options for the chunked GELF decoder.
+    #[serde(default, skip_serializing_if = "vector_core::serde::is_default")]
+    pub chunked_gelf: ChunkedGelfDecoderOptions,
+}
+
+impl ChunkedGelfDecoderConfig {
+    /// Build the `ChunkedGelfDecoder` from this configuration.
+    pub fn build(&self) -> ChunkedGelfDecoder {
+        ChunkedGelfDecoder::new(
+            self.chunked_gelf.timeout_millis,
+            self.chunked_gelf.pending_messages_limit,
+        )
+    }
+}
+
+/// Options for building a `ChunkedGelfDecoder`.
+#[configurable_component]
+#[derive(Clone, Debug, Derivative, PartialEq, Eq)]
+pub struct ChunkedGelfDecoderOptions {
+    /// The timeout, in milliseconds, for a message to be fully received. If the timeout is reached, the
+    /// decoder drops all the received chunks of the incomplete message and starts over.
+    /// The default value is 5 seconds.
+    #[serde(
+        default = "default_timeout_millis",
+        skip_serializing_if = "vector_core::serde::is_default"
+    )]
+    pub timeout_millis: u64,
+
+    /// The maximum number of pending incomplete messages. If this limit is reached, the decoder starts
+    /// dropping chunks of new messages. This limit ensures the memory usage of the decoder's state is bounded.
+    /// The default value is 1000.
+    #[serde(
+        default = "default_pending_messages_limit",
+        skip_serializing_if = "vector_core::serde::is_default"
+    )]
+    pub pending_messages_limit: usize,
+}
+
+impl Default for ChunkedGelfDecoderOptions {
+    fn default() -> Self {
+        Self {
+            timeout_millis: default_timeout_millis(),
+            pending_messages_limit: default_pending_messages_limit(),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct MessageState {
```

Comment:
I stumbled across https://docs.rs/timedmap/latest/timedmap/. It seems like that might be a fit here that would let us delete some code?

Reply:
Granted that seems to be a niche crate so it may not be worth the cost of adding another dependency.

Reply:
Agree with you!

Furthermore, I think that crate wouldn't allow to have a custom behaviour when evicting the key (i.e the `warn!` log)